linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/BinaryWriter.scala

```diff
@@ -0,0 +1,660 @@
+      vec(elemOpt.toList)(op)
+
+    def name(s: String): Unit =
+      name(UTF8String(s))
+
+    def name(utf8: UTF8String): Unit = {
+      val len = utf8.length
+      u32(len)
+      var i = 0
+      while (i != len) {
+        byte(utf8(i))
+        i += 1
+      }
+    }
+
+    def byteLengthSubSection(f: Buffer => Unit): Unit = {
+      // Reserve 4 bytes at the current offset to store the byteLength later
+      val byteLengthOffset = size
+      val startOffset = byteLengthOffset + 4
+      ensureCapacity(startOffset)
+      size = startOffset // do not write the 4 bytes for now
+
+      f(this)
+
+      // Compute byteLength
+      val endOffset = size
+      val byteLength = endOffset - startOffset
+
+      assert(byteLength < (1 << 28), s"Cannot write a subsection that large: $byteLength")
+
+      /* Write the byteLength in the reserved slot. Note that we *always* use
+       * 4 bytes to store the byteLength, even when less bytes are necessary in
+       * the unsigned LEB encoding. The WebAssembly spec specifically calls out
+       * this choice as valid. We leverage it to have predictable total offsets
+       * when write the code section, which is important to efficiently
+       * generate source maps.
+       */
+      buf(byteLengthOffset) = ((byteLength & 0x7F) | 0x80).toByte
+      buf(byteLengthOffset + 1) = (((byteLength >>> 7) & 0x7F) | 0x80).toByte
+      buf(byteLengthOffset + 2) = (((byteLength >>> 14) & 0x7F) | 0x80).toByte
+      buf(byteLengthOffset + 3) = ((byteLength >>> 21) & 0x7F).toByte
+    }
+
+    @tailrec
+    private def unsignedLEB128(value: Long): Unit = {
+      val next = value >>> 7
+      if (next == 0) {
+        byte(value.toByte)
+      } else {
+        byte(((value.toInt & 0x7F) | 0x80).toByte)
+        unsignedLEB128(next)
+      }
+    }
+
+    @tailrec
+    private def signedLEB128(value: Long): Unit = {
+      val chunk = value.toInt & 0x7F
+      val next = value >> 7
+      if (next == (if ((chunk & 0x40) != 0) -1 else 0)) {
+        byte(chunk.toByte)
+      } else {
+        byte((chunk | 0x80).toByte)
+        signedLEB128(next)
+      }
+    }
+  }
+
+  final class WithSourceMap(module: Module, emitDebugInfo: Boolean,
+      sourceMapWriter: SourceMapWriter, sourceMapURI: String)
+      extends BinaryWriter(module, emitDebugInfo) {
+
+    override protected def emitStartFuncPosition(buf: Buffer, pos: Position): Unit =
+      sourceMapWriter.startNode(buf.currentGlobalOffset, pos)
+
+    override protected def emitPosition(buf: Buffer, pos: Position): Unit = {
+      sourceMapWriter.endNode(buf.currentGlobalOffset)
+      sourceMapWriter.startNode(buf.currentGlobalOffset, pos)
+    }
+
+    override protected def emitEndFuncPosition(buf: Buffer): Unit =
+      sourceMapWriter.endNode(buf.currentGlobalOffset)
+
+    override protected def emitSourceMapSection(buf: Buffer): Unit = {
+      writeCustomSection(buf, "sourceMappingURL") { buf =>
```

Comment:
Ref this: https://github.com/WebAssembly/tool-conventions/blob/main/Debugging.md#source-maps ?